### PR TITLE
API Deprecations for template layer

### PR DESCRIPTION
**Pull request by GuySartorelli**
## Issue
- https://github.com/silverstripe/silverstripe-framework/issues/11322

### CODE REVIEW

File: src/Forms/GridField/GridFieldDataColumns.php
Comment (GuySartorelli) [inline src/Forms/GridField/GridFieldDataColumns.php:234]: Turns out this method just isn't used at all. It used to be used in `getColumnContent()` but the work this method used to do is now offloaded to `GridField::getDataFieldValue()`.

Found this because it's the only usage of `XML_val()` in core.

File: src/View/SSViewer.php
Comment (GuySartorelli) [inline src/View/SSViewer.php:92]: Whole bunch of stuff like this that's template-specific which is getting moved over to the new engine.

File: src/View/SSViewer.php
Comment (GuySartorelli) [inline src/View/SSViewer.php:141]: We don't really have a policy about what to do with properties - I've opted to deprecate it but didn't bother mentioning it in the docs

File: src/View/SSViewer.php
Comment (GuySartorelli) [inline src/View/SSViewer.php]: Literally nothing called this (except I think one place inside SSViewer itself)

File: src/View/SSViewer.php
Comment (GuySartorelli) [inline src/View/SSViewer.php:494]: Wasn't called anywhere - and ideally people will refer to templates more loosely as candidates rather than caring about the specific files. What are you ever going to do with a full template file path in code if you're not a template rendering engine?

File: src/View/SSViewer.php
Comment (GuySartorelli) [inline src/View/SSViewer.php:592]: Just a PHPDoc so doesn't break anything to change this - and I'm changing it to what is actually being passed in here.

File: src/View/SSViewer.php
Comment (GuySartorelli) [inline src/View/SSViewer.php:854]: Literally one place called this and it was a unit test.

File: src/View/SSViewer.php
Comment (GuySartorelli) [inline src/View/SSViewer.php:865]: Nothing called this so we clearly don't need to replace it in the new engine.

File: src/View/SSViewer.php
Comment (GuySartorelli) [inline src/View/SSViewer.php:884]: Renaming this, and passing in a boolean so it's easier to use by other engines.

But for now our template parser will keep calling this (hence why it's suppressed). I've opted to not add `Deprecation::withSuppressedNotice()` anywhere inside the template parser, so anything that calls has a suppressed notice directly in the method being called.

File: src/View/SSViewer.php
Comment (GuySartorelli) [inline src/View/SSViewer.php:898]: For simple methods like this it makes sense to implement in 5.4 so people can swap to it directly if they were calling it in their code.

File: src/View/SSViewer_DataPresenter.php
Comment (GuySartorelli) [inline src/View/SSViewer_DataPresenter.php:17]: I know this was intentionally kept separate, but whatever the case was when that decision was made it was causing me headaches trying to sort out what the difference is between `SSViewer_Scope` and this class. Especially since they have a lot of similar logic in them - the `__call()` class is a great example where the code flow is _not_ obvious with this class split out like this.

I've opted to merge the two classes together, and simplify where I can. There are or will be additional cards to simplify it further in the future, too.

File: src/View/SSViewer_FromString.php
Comment (GuySartorelli) [inline src/View/SSViewer_FromString.php:12]: Doesn't make sense to have this abstraction anymore - if you have a template string, it'll be for a specific template syntax so you should make sure to get the correct template engine from the injector and ask it to render your template string directly.

File: src/View/SSViewer_Scope.php
Comment (GuySartorelli) [inline src/View/SSViewer_Scope.php:131]: Just renamed it to be a little clearer that it's not _just_ getting the value of the `item` property. I ran into problems when I tried to replace all instances of `$this->item` with calling this method, because sometimes that's not what this method returns.

File: src/View/SSViewer_Scope.php
Comment (GuySartorelli) [inline src/View/SSViewer_Scope.php:211]: This is getting renamed so it's clear what it's for

File: src/View/ThemeResourceLoader.php
Comment (GuySartorelli) [inline src/View/ThemeResourceLoader.php:188]: Doesn't make sense for this to be here anymore, since this only knows about ss templates.

There is a replacement method for this in the new engine but I've opted to keep it private, to dissuade people from trying to deal with file paths for templates - we should be thinking and talking about template files as template "candidates" instead, i.e. "render with `Includes/Header`" and let the template engine figure out through theme inheritance which exact template file that maps to.

File: src/View/ViewableData.php
Comment (GuySartorelli) [inline src/View/ViewableData.php:431]: Nothing uses this, so no need to replace it. If we do need this in the future it can be added to the new `CastingService`.

File: src/View/ViewableData.php
Comment (GuySartorelli) [inline src/View/ViewableData.php:446]: Only one use of this in core, so no need to replace it. If we do need this in the future it can be added to the new `CastingService`.

File: src/View/ViewableData.php
Comment (GuySartorelli) [inline src/View/ViewableData.php:497]: You'll see in the implementations for `objCacheGet` and `objCacheSet` that they call this directly now, so there's no need for anyone else to be calling it.

File: src/View/ViewableData.php
Comment (GuySartorelli) [inline src/View/ViewableData.php]: That param was never used, and it's easier to reason about the cache if the cache name is always derived from the field name and arguments.

File: src/View/ViewableData.php
Comment (GuySartorelli) [inline src/View/ViewableData.php:602]: Nothing uses this, and if they did they should just call `obj()` directly instead.

File: src/View/ViewableData.php
Comment (GuySartorelli) [inline src/View/ViewableData.php:633]: This is currently called directly from `SSViewer_Scope` - that will be calling methods on `ViewLayerData` now instead, and the methods on `ViewLayerData` don't rely on this method, so there's no need to keep it.

File: src/View/ViewableData.php
Comment (GuySartorelli) [inline src/View/ViewableData.php:648]: Not sure why this exists, it's not used in sink.

**Review by emteknetnz — CHANGES_REQUESTED** [review]
I'm really not a fan of this pattern of wrapping calls to deprecated code with no replacement with Deprecation::withSuppressedNotice(), rather than just using it to wrap Deprecation::notice() in a central place

I can see why you've done it this way as it means that developers who do happen to call any of these methods will actually see the deprecation notices if they turn on deprecations. However they're pretty unlikely to be very useful as they're telling the developer to use a method that doesn't actually exist.

Wrapping calls means that there is FAR more code updates that need to be made, and we may still end up missing a call or two.  Just about every other CMS 5 PR could be closed if we didn't do it this way.

Looking at our own docs - https://docs.silverstripe.org/en/5/upgrading/deprecations/#enabling-deprecation-warnings - we do explicitly say that code with no replacement won't emit a deprecation notice by default, though there is a param available to turn on those notices too.

So if it's alright, could you please simply wrap the Deprecation::notice() with Deprecation::withSupressedNotice() and stop using wrapping the calls with withSupressedNotice()

**Comment by GuySartorelli** [comment]
Discussed offline - we're changing the way we do these a little bit. See https://github.com/silverstripe/silverstripe-framework/issues/11428

This is blocked until that gets merged

File: src/View/SSViewer.php
Comment (emteknetnz) [inline src/View/SSViewer.php]:
```suggestion
        Deprecation::noticeWithNoReplacment('5.4.0');
```

File: src/View/SSViewer.php
Comment (emteknetnz) [inline src/View/SSViewer.php]:
```suggestion
        Deprecation::noticeWithNoReplacment('5.4.0');
```

File: src/View/SSViewer.php
Comment (emteknetnz) [inline src/View/SSViewer.php]:
```suggestion
        Deprecation::noticeWithNoReplacment('5.4.0');
```

File: src/View/SSViewer.php
Comment (emteknetnz) [inline src/View/SSViewer.php:345]:
```suggestion
        Deprecation::noticeWithNoReplacment('5.4.0');
```

File: src/View/SSViewer.php
Comment (emteknetnz) [inline src/View/SSViewer.php]:
```suggestion
        Deprecation::noticeWithNoReplacment('5.4.0');
```

File: src/View/SSViewer.php
Comment (emteknetnz) [inline src/View/SSViewer.php]:
```suggestion
        Deprecation::noticeWithNoReplacment('5.4.0');
```

File: src/View/ThemeResourceLoader.php
Comment (emteknetnz) [inline src/View/ThemeResourceLoader.php:190]:
```suggestion
        Deprecation::noticeWithNoReplacment('5.4.0');
```

File: src/View/ViewableData.php
Comment (emteknetnz) [inline src/View/ViewableData.php:433]:
```suggestion
        Deprecation::noticeWithNoReplacment('5.4.0');
```

File: src/View/ViewableData.php
Comment (emteknetnz) [inline src/View/ViewableData.php:448]:
```suggestion
        Deprecation::noticeWithNoReplacment('5.4.0');
```

File: src/View/ViewableData.php
Comment (emteknetnz) [inline src/View/ViewableData.php]:
```suggestion
        Deprecation::noticeWithNoReplacment('5.4.0');
```

File: src/View/ViewableData.php
Comment (emteknetnz) [inline src/View/ViewableData.php]:
```suggestion
        Deprecation::noticeWithNoReplacment('5.4.0');
```

**Review by emteknetnz — CHANGES_REQUESTED** [review]

**Comment by GuySartorelli** [comment]
Dunno if I caught all the ones you listed - but as discussed in person I've done a very basic find and replace. If I missed any... well, it's the same message anyway so it's not worth updating them.

**Review by emteknetnz — APPROVED** [review]

**Pull request merged**
